diff --git a/audio/common/all-versions/default/service/Android.bp b/audio/common/all-versions/default/service/Android.bp
index 9890be295..5d5c2eef2 100644
--- a/audio/common/all-versions/default/service/Android.bp
+++ b/audio/common/all-versions/default/service/Android.bp
@@ -64,8 +64,43 @@ cc_binary {
     ],
 }
 
+cc_binary {
+    name: "android.hardware.audio.service_64",
+
+    init_rc: ["android.hardware.audio.service_64.rc"],
+    relative_install_path: "hw",
+    vendor: true,
+    compile_multilib: "64",
+    srcs: ["service.cpp"],
+
+    cflags: [
+        "-Wall",
+        "-Wextra",
+        "-Werror",
+    ],
+
+    shared_libs: [
+        "libcutils",
+        "libbinder",
+        "libbinder_ndk",
+        "libhidlbase",
+        "liblog",
+        "libutils",
+        "libhardware",
+    ],
+
+    defaults: [
+        "android_hardware_audio_config_defaults",
+    ],
+}
+
 // Legacy service name, use android.hardware.audio.service instead
 phony {
     name: "android.hardware.audio@2.0-service",
     required: ["android.hardware.audio.service"],
 }
+
+phony {
+    name: "android.hardware.audio@2.0-service_64",
+    required: ["android.hardware.audio.service_64"],
+}
\ No newline at end of file
diff --git a/audio/common/all-versions/default/service/android.hardware.audio.service_64.rc b/audio/common/all-versions/default/service/android.hardware.audio.service_64.rc
new file mode 100644
index 000000000..507cb347c
--- /dev/null
+++ b/audio/common/all-versions/default/service/android.hardware.audio.service_64.rc
@@ -0,0 +1,12 @@
+service vendor.audio-hal /vendor/bin/hw/android.hardware.audio.service_64
+    override
+    class hal
+    user audioserver
+    # media gid needed for /dev/fm (radio) and for /data/misc/media (tee)
+    group audio camera drmrpc inet media mediadrm net_bt net_bt_admin net_bw_acct oem_2901 wakelock
+    capabilities BLOCK_SUSPEND
+    ioprio rt 4
+    writepid /dev/cpuset/foreground/tasks /dev/stune/foreground/tasks
+    socket audio_us_socket_0 seqpacket 0666 system system
+    onrestart restart audioserver
+
